core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -300,36 +329,73 @@ private void tryStart() {
             CreateContainerCmd createCommand = dockerClient.createContainerCmd(dockerImageName);
             applyConfiguration(createCommand);
 
-            containerId = createCommand.exec().getId();
+            createCommand.getLabels().put(DockerClientFactory.TESTCONTAINERS_LABEL, "true");
 
-            connectToPortForwardingNetwork(createCommand.getNetworkMode());
+            if (shouldBeReused) {
+                if (!canBeReused()) {
+                    throw new IllegalStateException("This container does not support reuse");
+                }
+
+                if (TestcontainersConfiguration.getInstance().environmentSupportsReuse()) {
+                    String hash = hash(createCommand);
+
+                    containerId = findContainerForReuse(hash).orElse(null);
 
-            copyToFileContainerPathMap.forEach(this::copyFileToContainer);
+                    if (containerId != null) {
+                        logger().info("Reusing container with ID: {} and hash: {}", containerId, hash);
+                        containerInfo = dockerClient.inspectContainerCmd(containerId).exec();
+                    } else {
+                        logger().debug("Can't find a reusable running container with hash: {}", hash);
 
-            containerIsCreated(containerId);
+                        createCommand.getLabels().put(HASH_LABEL, hash);
+                    }
+                } else {
+                    logger().info("Reuse was requested but the environment does not support the reuse of containers");
+                }
+            } else {
+                createCommand.getLabels().put(DockerClientFactory.TESTCONTAINERS_SESSION_ID_LABEL, DockerClientFactory.SESSION_ID);
+            }
 
-            logger().info("Starting container with ID: {}", containerId);
-            dockerClient.startContainerCmd(containerId).exec();
+            if (containerInfo == null) {
```

Comment:
Important information for the reviewers:
these `containerInfo == null` branches are important to review carefully